packages/controllers/src/snaps/keyring.ts

```diff
@@ -0,0 +1,89 @@
+// TODO
+export interface Namespace {
+  chains: { id: string; name: string }[];
+  methods?: string[];
+  events?: string[];
+}
+
+/**
+ * Finds a keyring snap for each namespace implements at a minimum the requested functionality.
+ *
+ * @param requestedNamespaces - The requested namespaces.
+ * @param snaps - All snaps and their exposed keyring namespaces.
+ * @returns A mapping between namespaces and snap ids.
+ */
+export function findMatchingKeyringSnaps(
+  requestedNamespaces: Record<string, Namespace>,
+  snaps: Record<string, Record<string, Namespace>>,
+) {
+  const snapEntries = Object.entries(snaps);
+  return Object.entries(requestedNamespaces).reduce(
+    (acc, [requestedNamespaceId, currentRequestedNamespace]) => {
+      const matchedSnap = snapEntries.find(([_, namespaces]) => {
+        const potentialMatch = namespaces[requestedNamespaceId];
+        return (
+          potentialMatch &&
+          matchNamespace(currentRequestedNamespace, potentialMatch)
+        );
+      });
+      return { ...acc, [requestedNamespaceId]: matchedSnap?.[0] ?? null };
```

Comment:
Maybe to not clutter the object with possible `null` values:

```suggestion
      const value = matchedSnap?.[0];
      if (value) {
        return { ...acc, [requestedNamespaceId]: value };
      }
      
      return acc;
```

Reply:
I assumed we want to return `null` for non matches? It could just be undefined, but I wonder if its better to be explicit?

Reply:
Not sure, up to you!

Reply:
Lets keep it as explicit null for now, it's an easy change if we decide otherwise later.